sky/packages/sky/lib/src/widgets/navigator2.dart

```diff
@@ -68,139 +69,68 @@ class Navigator extends StatefulComponent {
-    return config.onGeneratePage(name);
+  bool get hasPreviousRoute => _modal.length > 1;
+  OverlayState get overlay => _overlayKey.currentState;
+
+  OverlayEntry get _currentOverlay {
+    for (Route route in _ephemeral.reversed) {
+      if (route.topEntry != null)
+        return route.topEntry;
+    }
+    for (Route route in _modal.reversed) {
+      if (route.topEntry != null)
+        return route.topEntry;
+    }
+    return null;
   }
 
-  bool get hasPreviousRoute => _history.length > 1;
+  Route get _currentRoute => _ephemeral.isNotEmpty ? _ephemeral.last : _modal.last;
+  Route _removeCurrentRoute() => _ephemeral.isNotEmpty ? _ephemeral.removeLast() : _modal.removeLast();
 
   void pushNamed(String name, { Set<Key> mostValuableKeys }) {
-    final RouteBuilder builder = config.routes[name] ?? _generatePage(name) ?? config.onUnknownPage;
-    assert(builder != null); // 404 getting your 404!
-    push(new PageRoute(
-      builder: builder,
-      name: name,
-      mostValuableKeys: mostValuableKeys
-    ));
-  }
-
-  void _addRouteToHistory(Route route) {
-    route.willPush();
-    route._entry = new OverlayEntry(child: route._child);
-    _history.add(route);
+    RouteArguments args = new RouteArguments(name: name, mostValuableKeys: mostValuableKeys);
+    push(config.onGenerateRoute(args) ?? config.onUnknownRoute(args));
   }
 
   void push(Route route) {
-    OverlayEntry reference = _history.last._entry;
-    _addRouteToHistory(route);
-    _overlay.currentState.insert(route._entry, above: reference);
+    _popAllEphemeralRoutes();
+    route.add(overlay, _currentOverlay);
+    _modal.add(route);
+    route.didMakeCurrent();
   }
 
-  void pop([dynamic result]) {
-    _history.removeLast().didPop(result);
-  }
-
-  Widget build(BuildContext context) {
-    return new Overlay(
-      key: _overlay,
-      initialEntries: <OverlayEntry>[ _history.first._entry ]
-    );
-  }
-}
-
-abstract class TransitionRoute extends Route {
-  PerformanceView get performance => _performance?.view;
-  Performance _performance;
-
-  Duration get transitionDuration;
-
-  Performance createPerformance() {
-    Duration duration = transitionDuration;
-    assert(duration != null && duration >= Duration.ZERO);
-    return new Performance(duration: duration, debugLabel: debugLabel);
+  void pushEphemeral(Route route) {
+    route.add(overlay, _currentOverlay);
+    _ephemeral.add(route);
+    route.didMakeCurrent();
   }
 
-  void willPush() {
-    _performance = createPerformance();
-    _performance.forward();
-    super.willPush();
+  void _popAllEphemeralRoutes() {
+    List<Route> localEphemeral = new List<Route>.from(_ephemeral);
+    _ephemeral.clear();
+    for (Route route in localEphemeral)
+      route.remove(null);
+    assert(_ephemeral.isEmpty);
   }
 
-  Future didPop(dynamic result) async {
-    await _performance.reverse();
-    super.didPop(result);
+  void pop([dynamic result]) {
+    _removeCurrentRoute().remove(result);
```

Comment:
We changed the old navigator not to do that.  This behavior is the same as the old navigator.